spanner/value.go

```diff
@@ -944,9 +1034,9 @@ func errNilArrElemType(t *sppb.Type) error {
 	return spannerErrorf(codes.FailedPrecondition, "array type %v is with nil array element type", t)
 }
 
-// errNilDst returns error for decoding into nil interface{}.
-func errNilSrcPtr(dst interface{}) error {
-	return spannerErrorf(codes.InvalidArgument, "cannot use nil type %T", dst)
+// errNotValidSrc returns error is valid field is false
+func errNotValidSrc(dst interface{}) error {
+	return spannerErrorf(codes.InvalidArgument, "cannot have valid field as false in %T", dst)
```

Comment:
I don't quite understand why we are changing this generic error? And also, the new error message does not really make sense to me. `cannot use nil type ...` would make me as a user understand that I used a nil where I should not. `cannot have valid field as false in ...` would not make me understand that I for example used a `nil` reference where I should not.

Reply:
@olavloite 
Previously we were restricting users from passing a custom nil type of type proto (ex: `(*pb.SingerInfo)(nil)`) due to the issue pointed out in [PR](https://github.com/googleapis/google-cloud-go/pull/6886#discussion_r1006305470). However, I found out a solution to allow nil types and therefore this error `errNilSrcPtr` is no longer needed. 
Note: Removing this error `errNilSrcPtr` will not affect existing code as no other methods use this except the Proto changes.

Now coming to error `errNotValidSrc`, custom types like `NullProto` and `NullEnum` have a field called `valid`. Users should explicitly set this field as `true` to indicate that the `value` in the struct is valid. However, if the `encodeValue` method receives this struct with `valid` as `false`, then the above error gets thrown to user.

A sample error would look like below
`cannot have valid field as false in spanner.NullProto`

Reply:
@harshachinta Thanks for the explanation.

I would still not know what I'm doing wrong if I get the error message `cannot have valid field as false in spanner.NullProto`. Is this field never allowed to be false? If so, why is it there? Or is it only disallowed in some cases? If so, the error message should indicate that.
Or is this some internal error that should never be returned, and is only there as a safety measure?

Reply:
> Is this field never allowed to be false?

yes @olavloite . This field is used explicitly during `read` operation. This field `valid` is set by client library to `false`, whenever the data from db is `null`. 
During `write` operation, especially for protos we do not allow users to set `valid` field to be false. Other types like `NullString`, `NullInt64` considers this as a `null` value. However, for Protos we need to derive `proto_type_fqn` from the input being passed, and so does not allow users to set it to `false`. Hence the above exception gets thrown `cannot have valid field as false in spanner.NullProto`

Reply:
If that's the case, consider making the error msg more explicit that users are not expected to use this for writing.

But it does make it inconsistent with other Null{Type} structs.

For the case of not being able to infer the fqn, does go client library have the concept of untyped null? Are we able to make use of untyped null?

Reply:
I would expect this error message in that case to:
1. Say something like `field "valid" of spanner.NullProto may not be set to false when writing data to Cloud Spanner`. I feel that that explains better to the user exactly what they are doing wrong.
2. Include a hint that tells the user how they should write a null value, as this is inconsistent with other `Null{Type}` structs.

Reply:
@olavloite 
As suggested I have modified the error to be more readable. Can you please review it?